datalad_next/utils/shell_connection.py

```diff
@@ -0,0 +1,305 @@
+    detect the end of the output and the determine the return code of the
+    executed command.
+
+    Both, command output and command return code are available via
+    :class:`ShellCommandResponseGenerator`-instances that are returned by the
+    :meth:`execute_command`-method of this class.
+    """
+
+    def __init__(self,
+                 process_inputs: Queue,
+                 stdout: OutputFrom,
+                 ) -> None:
+        self.process_inputs = process_inputs
+        self.stdout = stdout
+        self.end_marker, self.command_postfix = self._get_marker()
+
+    def execute_command(self,
+                        command: bytes,
+                        *,
+                        stdin: Iterable[bytes] | None = None,
+                        ) -> ShellCommandResponseGenerator:
+        """Execute a command in the connected shell
+
+        Parameters
+        ----------
+        command : bytes
+            The command to execute. Command can be a single POSIX pipeline list.
+            TODO: provide an example
+            (Internally the command is extended to mark the end of the output
+            by printing an end marker and the return code.)
+        stdin : Iterable[byte] | None, optional, default: None
+            If given, the bytes are sent to the stdin of the command.
+
+            If the
+            command reads its ``stdin`` until EOF, you have to use
+            :meth:`self.close` to close ``stdin`` of the command. Otherwise
+            the command will usually not terminate.
+            Once :meth:`self.close` is called, no more commands can be executed
+            with this :class:`ShellCommandExecutor`-instance.
+
+        Yields
+        ------
+        :class:`ShellCommandResponseGenerator`
+
+            A generator that yields the output of ``stdout`` of the command.
+            The generator is exhausted when all output is read. After that,
+            the return code of the command execution is available in the
+            ``code``-attribute of the generator.
+        """
+        self.process_inputs.put([command + self.command_postfix])
+        if stdin is not None:
+            self.process_inputs.put(stdin)
+        return ShellCommandResponseGenerator(self.stdout, self.end_marker)
+
+    def upload(self,
+               local_path: Path,
+               remote_path: PurePosixPath,
+               ) -> tuple[int, bytes]:
+        """Upload a local file to a named file in the connected shell
+
+        Parameters
+        ----------
+        local_path : Path
+            The file that should be uploaded.
+        remote_path : PurePosixPath
+            The name of the file on the connected shell that will contain the
+            uploaded content.
+
+        Returns
+        -------
+        tuple[int, bytes]
+            A tuple containing the return code of the command that was executed
+            to upload the content and the last 64k bytes of stderr output
```

Comment:
What is the rationale for "64k"? It is not set here, it just happens to be 64k because of the setup of the underlying tooling. Would be better to make that connection in the docs too.

Reply:
Good point, I forgot that `chunk_size` is exposed at `iter_subproc`. Will change it.

Reply:
done in 1026ae678db87c2dba729ec4598147d90fafdafe